Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp

```diff
@@ -199,6 +268,269 @@ namespace AZ
+                }
+
+                auto shaderFile = AZ::Utils::ReadFile<AZStd::string>(shaderTemplate.m_shader);
+                if (!shaderFile.IsSuccess())
+                {
+                    AZ_Error(MaterialTypeBuilderName, false, "Failed to load shader template file '%s'. %s", shaderTemplate.m_shader.c_str(), shaderFile.GetError().c_str());
+                    return;
+                }
+
+                // At this point shaderTemplate.m_azsli should be an absolute path due to ResolvePathReference() being called above.
+                // It might be better for the include path to be relative to the generated .shader file path in the intermediate cache,
+                // so the project could be renamed or moved without having to rebuild the cache. But there's a good chance that moving
+                // the project would require a rebuild of the cache anyway.
+                AZStd::string includeTemplateAzslPath = shaderTemplate.m_azsli;
+
+                // Intermediate azsl file
+
+                AZStd::string generatedAzsl = AZStd::string::format(
+                    "// This code was generated by %s. Do not modify.\n"
+                    "#include <%s>\n",
+                    MaterialTypeBuilderName,
+                    includeTemplateAzslPath.c_str());
+
+                // The exact same material-specific shader code is included in every shader
+                if (!materialAzsliFilePath.empty())
+                {
+                    generatedAzsl += AZStd::string::format("#include <%s>\n", materialAzsliFilePath.c_str());
+                }
+
+                // The "_DEFINED" macros allow the material type to provide code for only the functions it needs to customize,
+                // and all other functions will be stubbed out.
+                generatedAzsl +=
+                    " \n"
+                    "#if !MaterialFunction_AdjustVertexData_DEFINED                                                    \n"
+                    "    void MaterialFunction_AdjustVertexData(in float3 positionLS, inout VertexData vertexData) {}  \n"
+                    "    #define MaterialFunction_AdjustVertexData_DEFINED 1                                           \n"
+                    "#endif                                                                                            \n"
+                    "                                                                                                  \n"
+                    "#if !MaterialFunction_AdjustSurface_DEFINED && MATERIALPIPELINE_SHADER_HAS_PIXEL_STAGE            \n"
+                    "    void MaterialFunction_AdjustSurface(inout Surface outSurface) {}                              \n"
+                    "    #define MaterialFunction_AdjustSurface_DEFINED 1                                              \n"
+                    "#endif                                                                                            \n"
+                    ;
```

Comment:
Other than the include file, will the generated code ever changed dynamically? Could this be moved to a template file and copied over instead? Then it can be updated if necessary without recompiling the builder.

Reply:
Yes I have been thinking about moving this out. I didn't bother for now since I'll be iterating on this code a lot in the next couple weeks, I'll have lots of other opportunities to move this out. If I have to touch this snippet at all, I'll probably move it to a file at that time.